sdk/python/kfp/compiler/compiler.py

```diff
@@ -714,51 +714,87 @@ def _sanitize_and_inject_artifact(self, pipeline: dsl.Pipeline):
       sanitized_ops[sanitized_name] = op
     pipeline.ops = sanitized_ops
 
-  def _compile(self, pipeline_func):
-    """Compile the given pipeline function into workflow."""
-    # Step 1: extract param value, name and description from pipeline_func signature and decoration.
+  def create_workflow_from_func(self,
+      pipeline_func: Callable,
+      pipeline_name: Text,
+      pipeline_description: Text=None,
+      params_list: List[dsl.PipelineParam]=()) -> Dict[Text, Any]:
+    """ Create workflow spec from pipeline function and specified pipeline
+    params/metadata. Currently, the pipeline params are either specified in
+    the signature of the pipeline function or by passing a list of
+    dsl.PipelineParam. Conflict will cause ValueError.
+
+    :param pipeline_func: pipeline function where ContainerOps are invoked.
+    :param pipeline_name:
+    :param pipeline_description:
+    :param params_list: list of pipeline params to append to the pipeline.
+    :return: workflow dict.
+    """
     argspec = inspect.getfullargspec(pipeline_func)
 
     # Create the arg list with no default values and call pipeline function.
     # Assign type information to the PipelineParam
     pipeline_meta = _extract_pipeline_metadata(pipeline_func)
+    pipeline_meta.name = pipeline_name or pipeline_meta.name
+    pipeline_meta.description = pipeline_description or pipeline_meta.description
     pipeline_name = K8sHelper.sanitize_k8s_name(pipeline_meta.name)
 
+    # Currently only allow specifying pipeline params at one place.
+    if params_list and pipeline_meta.inputs:
+      raise ValueError('Only support specifying params by pipeline function signature, or params_list')
+
     args_list = []
-    for arg_name in argspec.args:
-      arg_type = None
-      for input in pipeline_meta.inputs:
-        if arg_name == input.name:
-          arg_type = input.param_type
-          break
-      args_list.append(dsl.PipelineParam(K8sHelper.sanitize_k8s_name(arg_name), param_type=arg_type))
-
-    # Step 2: Inflate pipeline obj with ContainerOps.
-    with dsl.Pipeline(pipeline_name) as p:
+    if pipeline_meta.inputs:
+      for arg_name in argspec.args:
```

Comment:
Use dict comprehension to avoid iterating/indentation twice:
```
if pipeline_meta.inputs:
  input_types = { input.name : input.param_type for input in pipeline_meta.inputs }
  for arg_name in argspec.args:
    arg_type = input_types.get(arg_name, None)
    arg_list.append(dsl.PipelineParam(K8sHelper.sanitize_k8s_name(arg_name), param_type=arg_type))
```

Reply:
Done. Thanks!